pennylane/allocation.py

```diff
@@ -0,0 +1,270 @@
+# limitations under the License.
+"""
+This module contains the commands for allocating and freeing wires dynamically.
+"""
+import uuid
+from typing import Optional, Sequence
+
+from pennylane.operation import Operator
+from pennylane.wires import Wires
+
+
+class DynamicWire:
+    """A wire whose concrete value will be determined later during a compilation step or execution.
+
+    Multiple dynamic wires can correspond to the same device wire as long as they are properly allocated and
+    freed.
+
+    Args:
+        key (Optional[str]): a ``uuid4`` string to uniquely identify the dynamic wire.
+    """
+
+    def __init__(self, key: Optional[uuid.UUID] = None):
+        self.key = key or uuid.uuid4()
+
+    def __eq__(self, other):
+        if not isinstance(other, DynamicWire):
+            return False
+        return self.key == other.key
+
+    def __hash__(self):
+        return hash(("DynamicWire", self.key))
+
+    def __repr__(self):
+        return "<DynamicWire>"
+
+
+class Allocate(Operator):
+    """An instruction to values for dynamic wires.
+
+    Args:
+        wires (list[DynamicWire]): a list of dynamic wire values.
+
+    Keyword Args:
+        require_zeros (bool): Whether or not the wire must start in a ``0`` state.
+
+    ..see-also:: :func:`~.allocate`, :func:`~.safe_allocate`.
+
+    """
+
+    def __init__(self, wires, require_zeros=True):
+        super().__init__(wires=wires)
+        self._hyperparameters = {"require_zeros": require_zeros}
+
+    @property
+    def require_zeros(self):
+        """Whether or not the allocated wires are required to be in the zero state."""
+        return self.hyperparameters["require_zeros"]
+
+    @classmethod
+    def from_num_wires(cls, num_wires: int, require_zeros=True) -> "Allocate":
+        """Initialize an ``Allocate`` op from a number of wires instead of already constructed dynamic wires."""
+        wires = tuple(DynamicWire() for _ in range(num_wires))
+        return cls(wires=wires, require_zeros=require_zeros)
+
+
+class Deallocate(Operator):
+    """An instruction to deallocate the provided ``DynamicWire``'s.
+
+    Args:
+        wires (DynamicWire, Sequence[DynamicWire]): one or more dynamic wires to deallocate.
+        reset_to_original (bool): Whether or not the qubit will be in the same state upon being freed.
+
+    """
+
+    def __init__(self, wires: DynamicWire | Sequence[DynamicWire], reset_to_original=False):
+        super().__init__(wires=wires)
+        self._hyperparameters = {"reset_to_original": reset_to_original}
+
+    @property
+    def reset_to_original(self):
+        """Whether or not the dynamic wire will be returned to its original state."""
+        return self.hyperparameters["reset_to_original"]
+
+
+def allocate(num_wires: int, require_zeros: bool = True) -> Wires:
+    """Allocate dynamic wires for temporary use within a circuit.
+
+    .. warning::
+
+        This feature is experimental and may not be possible on all devices.
+
```

Comment:
👀

Reply:
Yeah... I'll update to `any` devices for now.